design/apidsl/security.go

```diff
@@ -0,0 +1,308 @@
+// "basic".
+//
+// Example:
+//
+//     OtherSecurity("jwt", "apiKey", func() {
+//	    	InHeader("Authorization")
+//	    	AuthorizationURL("http://example.com/authorization")
+//	    	Scope("my_system:write", "Write to the system")
+//	    	Scope("my_system:read", "Read anything in there")
+//     })
+//
+func OtherSecurity(name string, securityType string, dsl ...func()) {
+	if _, ok := dslengine.CurrentDefinition().(*design.APIDefinition); !ok {
+		dslengine.IncompatibleDSL()
+		return
+	}
+
+	if securityMethodRedefined(name) {
+		return
+	}
+
+	def := &design.SecurityMethodDefinition{
+		Kind:   design.OtherSecurityKind,
+		Method: name,
+		Type:   securityType,
+	}
+
+	if len(dsl) != 0 {
+		if !dslengine.Execute(dsl[0], def) {
+			return
+		}
+	}
+
+	design.Design.SecurityMethods = append(design.Design.SecurityMethods, def)
+}
+
+// Scope defines an authorization scope. Used within SecurityMethod,
+// the description is required, explaining what the scope
+// means. Within a Security block, only a scope is needed.
+func Scope(name string, desc ...string) {
+	switch parent := dslengine.CurrentDefinition().(type) {
+	case *design.SecurityDefinition:
+		if len(desc) == 1 {
+			dslengine.ReportError("must not pass description to Scope()")
+			return
+		}
+		parent.Scopes = append(parent.Scopes, name)
+	case *design.SecurityMethodDefinition:
+		if len(desc) == 0 {
+			dslengine.ReportError("no description provided for Scope()")
+			return
+		}
+		if parent.Scopes == nil {
+			parent.Scopes = make(map[string]string)
+		}
+		parent.Scopes[name] = desc[0]
+	default:
+		dslengine.IncompatibleDSL()
+	}
+}
+
+// InHeader defines both `in` and `name` of a SecurityMethod.  See
+// http://swagger.io/specification/#securitySchemeObject for details.
+func InHeader(headerName string) {
+	if parent, ok := dslengine.CurrentDefinition().(*design.SecurityMethodDefinition); ok {
+		if parent.Kind == design.APIKeySecurityKind || parent.Kind == design.OtherSecurityKind {
+			if parent.In != "" {
+				dslengine.ReportError("'In' previously defined through InHeader or InQuery")
+				return
+			}
+			parent.In = "header"
+			parent.Name = headerName
+			return
+		}
+	}
+	dslengine.IncompatibleDSL()
+}
+
+// InQuery defines both `in` and `name` of a SecurityMethod. See
+// http://swagger.io/specification/#securitySchemeObject for details.
+func InQuery(parameterName string) {
+	if parent, ok := dslengine.CurrentDefinition().(*design.SecurityMethodDefinition); ok {
+		if parent.Kind == design.APIKeySecurityKind || parent.Kind == design.OtherSecurityKind {
+			if parent.In != "" {
+				dslengine.ReportError("'In' previously defined through InHeader or InQuery")
+				return
+			}
+			parent.In = "query"
+			parent.Name = parameterName
+			return
+		}
+	}
+	dslengine.IncompatibleDSL()
+}
+
+// OAuth2Flow defines the `flow` for a SecurityMethod.  See
+// http://swagger.io/specification/#securitySchemeObject for details.
```

Comment:
As a high level comment, I am not sure about referring to the swagger spec in the DSL. The DSL does not currently rely explicitly on the Swagger specification and that seems to make sense to me.  I see the goa DSL as a super set and Swagger being just one way to "project" it. With that model it seems better to me to define the DSL independently of Swagger. Maybe that doesn't change anything to the DSL and only the comments or it could actually mean making the DSL more independent / meaningful on its own which wouldn't be a bad thing.